File: Sources/ClipDeckKit/Model/LibraryStorage.swift
import Foundation

/// Leitura e escrita da biblioteca em disco.
///
/// Separado do `Store` de propósito: aqui não há Combine nem estado observável,
/// então dá para testar escrita atômica, backup e recuperação sem levantar o app.
public struct LibraryStorage {
    public enum LoadOutcome: Equatable {
        case loaded
        case seededNewFile
        case recoveredFromBackup(String)
        case seededAfterCorruption(String)
    }

    public let directory: URL
    private let maxBackups = 5

    public init(directory: URL) {
        self.directory = directory
    }

    public static var defaultDirectory: URL {
        FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("ClipDeck", isDirectory: true)
    }

    public var fileURL: URL { directory.appendingPathComponent("library.json") }
    public var backupsDirectory: URL { directory.appendingPathComponent("Backups", isDirectory: true) }

    // JSON legível e estável: dá para versionar no git, sincronizar via iCloud
    // Drive e editar à mão sem o diff explodir a cada gravação.
    private static var encoder: JSONEncoder {
        let e = JSONEncoder()
        e.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
        e.dateEncodingStrategy = .iso8601
        return e
    }

    private static var decoder: JSONDecoder {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .iso8601
        return d
    }

    /// Carrega a biblioteca, degradando em vez de falhar.
    ///
    /// Sem arquivo, semeia. Com arquivo corrompido, tenta o backup mais recente
    /// antes de desistir — e nunca apaga o arquivo ruim: renomeia para `.corrupt`
    /// para você poder resgatar o conteúdo à mão.
    public func load() throws -> (library: Library, outcome: LoadOutcome) {
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)

        guard FileManager.default.fileExists(atPath: fileURL.path) else {
            let seeded = Library.seeded()
            try write(seeded)
            return (seeded, .seededNewFile)
        }

        do {
            let data = try Data(contentsOf: fileURL)
            return (try Self.decoder.decode(Library.self, from: data), .loaded)
        } catch {
            let quarantine = directory.appendingPathComponent(
                "library.corrupt-\(Self.timestamp()).json"
            )
            try? FileManager.default.moveItem(at: fileURL, to: quarantine)

            if let (recovered, name) = try? mostRecentBackup() {
                try write(recovered)
                return (recovered, .recoveredFromBackup(name))
            }

            let seeded = Library.seeded()
            try write(seeded)
            return (seeded, .seededAfterCorruption(quarantine.lastPathComponent))
        }
    }

    /// Grava de forma atômica, com backup do estado anterior.
    public func save(_ library: Library) throws {
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        try backupCurrentFile()
        try write(library)
    }

    private func write(_ library: Library) throws {
        let data = try Self.encoder.encode(library)
        // .atomic grava num temporário e faz rename — uma queda no meio da
        // escrita deixa o arquivo antigo intacto em vez de meio arquivo.
        try data.write(to: fileURL, options: .atomic)
    }

    private func backupCurrentFile() throws {
        guard FileManager.default.fileExists(atPath: fileURL.path) else { return }
        try FileManager.default.createDirectory(at: backupsDirectory, withIntermediateDirectories: true)

        let destination = backupsDirectory
            .appendingPathComponent("library-\(Self.timestamp()).json")
        try? FileManager.default.copyItem(at: fileURL, to: destination)

        try pruneBackups()
    }

    private func pruneBackups() throws {
        let files = (try? FileManager.default.contentsOfDirectory(
            at: backupsDirectory,
            includingPropertiesForKeys: nil
        )) ?? []

        let sorted = files
            .filter { $0.lastPathComponent.hasPrefix("library-") }
            .sorted { $0.lastPathComponent > $1.lastPathComponent }

        for stale in sorted.dropFirst(maxBackups) {
            try? FileManager.default.removeItem(at: stale)
        }
    }

    private func mostRecentBackup() throws -> (Library, String)? {
        let files = (try? FileManager.default.contentsOfDirectory(
            at: backupsDirectory,
            includingPropertiesForKeys: nil
        )) ?? []

        let candidates = files
            .filter { $0.lastPathComponent.hasPrefix("library-") }
            .sorted { $0.lastPathComponent > $1.lastPathComponent }

        for candidate in candidates {
            if let data = try? Data(contentsOf: candidate),
               let library = try? Self.decoder.decode(Library.self, from: data) {
                return (library, candidate.lastPathComponent)
            }
        }
        return nil
    }

    /// Timestamp ordenável lexicograficamente — a ordenação por nome de arquivo
    /// nas rotinas de backup depende disso.
    private static func timestamp() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd-HHmmss-SSS"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(identifier: "UTC")
        return formatter.string(from: Date())
    }
}

File: windows/src/ClipDeck.Core/Models/LibraryStorage.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ClipDeck.Core.Models;

public enum LoadOutcomeKind { Loaded, SeededNewFile, RecoveredFromBackup, SeededAfterCorruption }

public readonly record struct LoadOutcome(LoadOutcomeKind Kind, string? Detail = null);

public readonly record struct LoadResult(Library Library, LoadOutcome Outcome);

/// <summary>
/// Leitura e escrita da biblioteca em disco. Sem estado observável, para poder
/// testar escrita atômica, backup e recuperação sem levantar o app.
/// </summary>
public sealed class LibraryStorage
{
    private const int MaxBackups = 5;

    public string Directory { get; }

    public LibraryStorage(string directory) => Directory = directory;

    /// <summary>
    /// %APPDATA%\ClipDeck no Windows; ~/.config/ClipDeck no macOS e Linux.
    /// </summary>
    public static string DefaultDirectory => Path.Combine(
        Environment.GetFolderPath(
            Environment.SpecialFolder.ApplicationData,
            Environment.SpecialFolderOption.Create),
        "ClipDeck");

    public string FilePath => Path.Combine(Directory, "library.json");
    public string BackupsDirectory => Path.Combine(Directory, "Backups");

    // JSON legível e estável: dá para versionar, sincronizar e editar à mão sem
    // o diff explodir a cada gravação.
    private static readonly JsonSerializerOptions Options = new()
    {
        WriteIndented = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true,
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    /// <summary>
    /// Carrega a biblioteca, degradando em vez de falhar. Sem arquivo, semeia.
    /// Com arquivo corrompido, tenta o backup mais recente antes de desistir — e
    /// nunca apaga o arquivo ruim: renomeia para .corrupt para resgate manual.
    /// </summary>
    public LoadResult Load()
    {
        System.IO.Directory.CreateDirectory(Directory);

        if (!File.Exists(FilePath))
        {
            var seeded = Library.Seeded();
            Write(seeded);
            return new LoadResult(seeded, new LoadOutcome(LoadOutcomeKind.SeededNewFile));
        }

        try
        {
            var json = File.ReadAllText(FilePath);
            var library = JsonSerializer.Deserialize<Library>(json, Options)
                          ?? throw new JsonException("conteúdo nulo");
            return new LoadResult(library, new LoadOutcome(LoadOutcomeKind.Loaded));
        }
        catch (Exception)
        {
            var quarantine = Path.Combine(Directory, $"library.corrupt-{Timestamp()}.json");
            try { File.Move(FilePath, quarantine); } catch { /* melhor esforço */ }

            var recovered = MostRecentBackup();
            if (recovered is { } found)
            {
                Write(found.Library);
                return new LoadResult(found.Library,
                    new LoadOutcome(LoadOutcomeKind.RecoveredFromBackup, found.Name));
            }

            var seeded = Library.Seeded();
            Write(seeded);
            return new LoadResult(seeded,
                new LoadOutcome(LoadOutcomeKind.SeededAfterCorruption, Path.GetFileName(quarantine)));
        }
    }

    /// <summary>Grava de forma atômica, com backup do estado anterior.</summary>
    public void Save(Library library)
    {
        System.IO.Directory.CreateDirectory(Directory);
        BackupCurrentFile();
        Write(library);
    }

    private void Write(Library library)
    {
        var json = JsonSerializer.Serialize(library, Options);

        // Grava num temporário e substitui: uma queda no meio da escrita deixa o
        // arquivo antigo intacto em vez de meio arquivo.
        var temp = FilePath + ".tmp";
        File.WriteAllText(temp, json);

        if (File.Exists(FilePath))
        {
            File.Replace(temp, FilePath, destinationBackupFileName: null);
        }
        else
        {
            File.Move(temp, FilePath);
        }
    }

    private void BackupCurrentFile()
    {
        if (!File.Exists(FilePath)) return;

        System.IO.Directory.CreateDirectory(BackupsDirectory);
        var destination = Path.Combine(BackupsDirectory, $"library-{Timestamp()}.json");
        try { File.Copy(FilePath, destination, overwrite: true); } catch { return; }

        PruneBackups();
    }

    private void PruneBackups()
    {
        if (!System.IO.Directory.Exists(BackupsDirectory)) return;

        var stale = System.IO.Directory
            .GetFiles(BackupsDirectory, "library-*.json")
            .OrderByDescending(Path.GetFileName, StringComparer.Ordinal)
            .Skip(MaxBackups);

        foreach (var file in stale)
        {
            try { File.Delete(file); } catch { /* melhor esforço */ }
        }
    }

    private (Library Library, string Name)? MostRecentBackup()
    {
        if (!System.IO.Directory.Exists(BackupsDirectory)) return null;

        var candidates = System.IO.Directory
            .GetFiles(BackupsDirectory, "library-*.json")
            .OrderByDescending(Path.GetFileName, StringComparer.Ordinal);

        foreach (var candidate in candidates)
        {
            try
            {
                var library = JsonSerializer.Deserialize<Library>(File.ReadAllText(candidate), Options);
                if (library is not null) return (library, Path.GetFileName(candidate));
            }
            catch { /* tenta o próximo */ }
        }

        return null;
    }

    /// <summary>
    /// Timestamp ordenável lexicograficamente — a rotação de backups ordena por
    /// nome de arquivo e depende disso.
    /// </summary>
    private static string Timestamp() =>
        DateTime.UtcNow.ToString("yyyyMMdd-HHmmss-fff", System.Globalization.CultureInfo.InvariantCulture);
}
